compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/SkikoUIViewDelegate.kt

```diff
@@ -0,0 +1,154 @@
+
+    @Suppress("DEPRECATION")
+    override fun pointInside(point: CValue<CGPoint>, event: UIEvent?): Boolean =
+        point.useContents {
+            val position = Offset(
+                (x * density.density).toFloat(),
+                (y * density.density).toFloat()
+            )
+
+            !scene.hitTestInteropView(position)
+        }
+
+    override fun onTouchesEvent(view: UIView, event: UIEvent, phase: UITouchesEventPhase) {
+        scene.sendPointerEvent(
+            eventType = phase.toPointerEventType(),
+            pointers = event.touchesForView(view)?.map {
+                val touch = it as UITouch
+                val id = touch.hashCode().toLong()
+                val position = touch.offsetInView(view, density.density)
+                ComposeScenePointer(
+                    id = PointerId(id),
+                    position = position,
+                    pressed = touch.isPressed,
+                    type = PointerType.Touch,
+                    pressure = touch.force.toFloat(),
+                    historical = event.historicalChangesForTouch(touch, view, density.density)
+                )
+            } ?: emptyList(),
+            timeMillis = (event.timestamp * 1e3).toLong(),
+            nativeEvent = event
+        )
+    }
+
+    override fun retrieveInteropTransaction(): UIKitInteropTransaction =
+        interopContext.retrieve()
+
+    override fun render(canvas: Canvas, targetTimestamp: NSTimeInterval) {
+        // The calculation is split in two instead of
+        // `(targetTimestamp * 1e9).toLong()`
+        // to avoid losing precision for fractional part
+        val integral = floor(targetTimestamp)
+        val fractional = targetTimestamp - integral
+        val secondsToNanos = 1_000_000_000L
+        val nanos = integral.roundToLong() * secondsToNanos + (fractional * 1e9).roundToLong()
+        val composeCanvas = canvas.asComposeCanvas()
+        val dx = metalOffset.x
+        val dy = metalOffset.y
+        composeCanvas.translate(dx, dy)
+        scene.render(composeCanvas, nanos)
+        composeCanvas.translate(-dx, -dy)
+    }
+
+}
+
+private fun UITouchesEventPhase.toPointerEventType(): PointerEventType =
+    when (this) {
+        UITouchesEventPhase.BEGAN -> PointerEventType.Press
+        UITouchesEventPhase.MOVED -> PointerEventType.Move
+        UITouchesEventPhase.ENDED -> PointerEventType.Release
+        UITouchesEventPhase.CANCELLED -> PointerEventType.Release
+    }
+
+private fun UIEvent.historicalChangesForTouch(
+    touch: UITouch,
+    view: UIView,
+    density: Float
+): List<HistoricalChange> {
+    val touches = coalescedTouchesForTouch(touch) ?: return emptyList()
+
+    return if (touches.size > 1) {
+        // subList last index is exclusive, so the last touch in the list is not included
+        // because it's the actual touch for which coalesced touches were requested
+        touches.subList(0, touches.size - 1).map {
```

Comment:
dropLast?

Reply:
I just copied this code.
Yeah looks like dropLast(1)